src/amo/middleware/prefixMiddleware.js

```diff
@@ -16,105 +16,96 @@ export function prefixMiddleware(req, res, next, { _config = config } = {}) {
+  if (hasValidLang && hasValidClientAppInPartTwo) {
+    log.debug('URL already has a valid lang and clientApp, nothing to fix');
+  } else if (hasValidLang) {
+    log.debug(`Prepending clientApp to URL: ${userAgentApp}`);
+    URLPathParts.splice(1, 0, userAgentApp);
     isApplicationFromHeader = true;
-    prependedOrMovedApplication = true;
-  } else if (!hasValidLang) {
-    // If lang wasn't valid or was missing prepend one.
+  } else if (hasValidClientAppInPartOne) {
     log.debug(`Prepending lang to URL: ${lang}`);
     URLPathParts.splice(0, 0, lang);
-    // If we've prepended the lang to the URL we need to re-check our
-    // URL exception and make sure it's valid.
-    hasValidClientAppUrlException = isValidClientAppUrlException(
-      URLPathParts[1],
-      {
-        _config,
-      },
-    );
+  } else if (
+    hasValidClientAppInPartTwo ||
+    hasValidClientAppUrlExceptionInPartTwo
+  ) {
+    log.debug(`Replacing lang in URL: ${URLPathParts[0]} with ${lang}`);
+    URLPathParts.splice(0, 1, lang);
+  } else {
+    log.debug(`Prepending lang and clientApp to URL: ${lang}/${userAgentApp}`);
+    URLPathParts.splice(0, 0, lang, userAgentApp);
+    isApplicationFromHeader = true;
   }
 
-  if (!hasValidClientApp && isValidClientApp(URLPathParts[1], { _config })) {
-    // We skip prepending an app if we'd previously prepended a lang and the
-    // 2nd part of the URL is now a valid app.
-    log.debug('Application in URL is valid following prepending a lang.');
-  } else if (prependedOrMovedApplication) {
-    log.debug(
-      'URL is valid because we added/changed the first part to a clientApp.',
-    );
-  } else if (hasValidLocaleException || hasValidClientAppUrlException) {
-    if (
-      clientAppRoutes.includes(URLPathParts[1]) === false &&
-      (hasValidLang || hasValidLocaleException)
-    ) {
-      log.debug('Exception in URL found; we fallback to addons-server.');
-      // TODO: Remove this once upgraded to react-router 4.
-      res.status(404).end(oneLine`This page does not exist in addons-frontend.
-        Returning 404; this error should trigger upstream (usually
-        addons-server) to return a valid response.`);
-      return next();
-    }
-    log.debug('Exception in URL found; prepending lang to URL.');
-  } else if (!hasValidClientApp) {
-    // If the app supplied is not valid we need to prepend one.
-    log.debug(`Prepending application to URL: ${application}`);
-    URLPathParts.splice(1, 0, application);
-    isApplicationFromHeader = true;
+  const hasValidLocaleException = isValidLocaleUrlException(URLPathParts[2], {
+    _config,
+  });
+  const hasValidClientAppUrlException = isValidClientAppUrlException(
+    URLPathParts[2],
+    {
+      _config,
+    },
+  );
+  if (hasValidClientAppUrlException) {
+    isApplicationFromHeader = false;
+  }
+
+  // clientApp values that are allowed through to the router
+  // TODO: This can be removed when we upgrade to react-router v4.
```

Comment:
We are on `react-router v4`, so I wonder if we can remove this now. As we are re-writing this whole module, it makes sense to deal with this now.

Reply:
What the comment didn't say: RRv4 added support for regular expressions in `Route` paths. If we can update the different paths (in https://github.com/mozilla/addons-frontend/blob/364dafc88412c3527cd49c9345404ccb64c68e20/src/amo/components/Routes/index.js), we can probably remove this now.

I think we'd need something like `:application(firefox|android)` instead of `:application` in _all_ paths (but I didn't read the docs).

Reply:
I'm not sure what changes are being requested here.

Reply:
Update the `path` values in each `Route` where there is `:application` so that we can remove `clientAppRoutes` in config and everywhere else (because `path` can be a regex now). And I think we should replace `:application` with `:application(firefox|android)` (untested, though).

Reply:
changing Routes/index.js seems a larger change (or at least an equal size change) than this patch.  If it's not required we should do it as a separate issue.

Reply:
It's essentially simplifying this middleware, which is rewritten here, so that's why we think it could be part of this patch.

Reply:
sure, if it was just a change to this chunk of code I'd agree with you, but it's not.

Reply:
FWIW I'm fine with it being a follow-up, as it does involve changing a file that is outside the scope of the current patch. Plus, those comments have been around for 4 years, and we upgraded to `react-router` ages ago, so there's no real urgency here. Even without addressing those `TODO`s, this is a big improvement over the previous `prefixMiddleware`.